### DISCUSSION TITLE
HammerDB. How to generate a compressible data?

### DISCUSSION THREAD

**Discussion by vbienf** [TPC-Council/HammerDB#507] [General]
Hi all! 

I didn't find any information how to generate compressible data with HammerDB. F.e. for sysbench I using for each string specific mask:

- sysbench.rand.string("name-@@@@@"), 
- sysbench.rand.string("street1-@@@@@@@@@@"),
- sysbench.rand.string("street2-@@@@@@@@@@"), 
- sysbench.rand.string("city-@@@@@@@@@@"), 
- etc...

link to source - https://github.com/Percona-Lab/sysbench-tpcc/blob/master/tpcc_common.lua

This mask helps to generate compressible data for my test.

How I can add a specific mask to string to generate data with HammerDB?

**Comment by sm-shaw** [comment]
Hi, Yes, we follow the TPC-C specification for random string generation, but if I understand the question you are saying that the default random strings from the specification do not compress very well when using solutions to compress the table data because they are well ... random and in some cases non-uniform. 

So the question is, is it possible to make the generated data less random, so it will compress better.  It is interesting that sysbench does this. It is not something we plan to do, however it is all open source, so you can change the generation of data as you wish.  If you look in the modules directory you can find the common data generation functions in `tpcccommon-1.0`

Using the example below for generating the address as you use, here is the function that does that: 

```
  #ADDRESS
  proc MakeAddress { chArray chalen } {
    return [ list [ MakeAlphaString 10 20 $chArray $chalen ] [ MakeAlphaString 10 20 $chArray $chalen ] [ MakeAlphaString 10 20 $chArray $chalen ] [ MakeAlphaString 2 2 $chArray $chalen ] [ Makezip ] ]
  }
```
and make address gets called here in the build script: 
```
proc District { mysql_handler w_id DIST_PER_WARE } {
    set globArray [ list 0 1 2 3 4 5 6 7 8 9 A B C D E F G H I J K L M N O P Q R S T U V W X Y Z a b c d e f g h i j k l m n o p q r s t u v w x y z ]
    set chalen [ llength $globArray ]
    puts "Loading District"
    set d_w_id $w_id
    set d_ytd 30000.0
    set d_next_o_id 3001
    for {set d_id 1} {$d_id <= $DIST_PER_WARE } {incr d_id } {
        set d_name [ MakeAlphaString 6 10 $globArray $chalen ]
        set d_add [ MakeAddress $globArray $chalen ]
        set d_tax_ran [ RandomNumber 10 20 ]
        set d_tax [ string replace [ format "%.2f" [ expr {$d_tax_ran / 100.0} ] ] 0 0 "" ]
        mysql::exec $mysql_handler "insert into district (`d_id`, `d_w_id`, `d_name`, `d_street_1`, `d_street_2`, `d_city`, `d_state`, `d_zip`, `d_tax`, `d_ytd`, `d_next_o_id`) values ('$d_id', '$d_w_id', '$d_name', '[ lindex $d_add 0 ]', '[ lindex $d_add 1 ]', '[ lindex $d_add 2 ]', '[ lindex $d_add 3 ]', '[ lindex $d_add 4 ]', '$d_tax', '$d_ytd', '$d_next_o_id')"
    }
    mysql::commit $mysql_handler 
    puts "District done"
    return
}
```
So for a similar effect you could either change Make Address or the build functions to prepend the random data with name-, street1- etc. 

Or for a really simple case of making things non-random change the character array (globArray) to have less options to select random data from, e.g. instead of all the alphabet, reduce the available options to fewer letters.

**Reply by vbienf** [reply]
Thank you, @sm-shaw for fast answer!

Yes, you understood it right, HammerDB is very good at generating random data and it doesn't correlate with the real world, where you have a too many identical pieces of data. Usually, this is not paid much attention to when testing, but in my case I'm testing a product that uses the data compression feature.

I'm not familiar with this tcl-language, but if I will change for this case:

from:
`set d_add [ MakeAddress $globArray $chalen ]`
to:
`set d_add [ list [street1-] MakeAddress $globArray $chalen ]`

it will work?

Also, do I need to recompile the hammerdb after this changes or is it enough to change `tpcccommon-1.0.tm` and `oraolap.tcl` files?

**Reply by sm-shaw** [reply]
Hi, 

It's easier than that, we have the string replace command that does what you need. So for example to overwrite the street and city strings with street1-, street2-, and city- you can do something like this:

```
% puts [ string replace "myrandomstring" 0 8 "street1-" ]
street1-tring
%
```
so string replace original-string start and end and new string

doing this you would change the module as follows: 

```
 proc MakeAddress { chArray chalen } {
#    return [ list [ MakeAlphaString 10 20 $chArray $chalen ] [ MakeAlphaString 10 20 $chArray $chalen ] [ MakeAlphaString 10 20 $chArray $chalen ] [ MakeAlphaString 2 2 $chArray $chalen ] [ Makezip ] ]
    return [ list [ string replace [ MakeAlphaString 10 20 $chArray $chalen ] 0 8 "street1-" ]  [ string replace [ MakeAlphaString 10 20 $chArray $chalen ] 0 8 "street2-" ] [ string replace [ MakeAlphaString 10 20 $chArray $chalen ] 0 5 "city-" ] [ MakeAlphaString 2 2 $chArray $chalen ] [ Makezip ] ]
  }

This would give output as follows for all addresses in customer and district for all databases and make it less random e.g.
```
![image](https://user-images.githubusercontent.com/38044085/217294108-99d3eced-0dc9-44f2-ba2a-2d05fd80c5f9.png)

No you don't need to compile anything, it all gets compiled into bytecode at runtime so just change the source. 

Still interesting why sysbench would see it necessary to do something like this by default thought.

**Reply by vbienf** [reply]
Nice! Thank you very much @sm-shaw for your answers!

According to this link https://www.percona.com/blog/tpcc-like-workload-sysbench-1-0/  they `want to be able to evaluate different compression methods in InnoDB and MyRocks.` This is a different DB engines for MySQL, that supports compression.

I will try to change scripts `tpcccommon-1.0.tm` and `oraolap.tcl` and add possibility to create data generation with different compression (2 and 3) and upload it to my repository. If someone need to generate compressible data with hammerDB they can just download that files and change them in main directory locally.

**Reply by sm-shaw** [reply]
One minor clarifying point - it will be tpcccommon-1.0.tm and oraoltp.tcl for TPROC-C and tpchcommon-1.0.tm and oraolap.tcl for TPROC-H. 
Note that changing the common module as would expected affects all databases.

**Reply by vbienf** [reply]
I understand that. But also in `oraoltp.tcl` there is a insert queries such us:

`proc mk_region { lda } {
    set sql INSERT INTO REGION (R_REGIONKEY, R_NAME, R_COMMENT) VALUES (:R_REGIONKEY, :R_NAME, :R_COMMENT)
    .....
        set comment [ TEXT_1 72 ]` 

I didn't find information about "TEXT_1" in `tpcccommon-1.0.tm`. So, I need to change not only common, but also all tcl files for all databases? Or how I can do that in rights way?

**Reply by vbienf** [reply]
@sm-shaw , oh, sorry! I found  "TEXT_1" in `tpchcommon-1.0.tm`. 

So, in this case I just need to change both files `tpcccommon-1.0.tm` and `tpchcommon-1.0.tm` and it will affect to all databases?

**Comment by sm-shaw** [comment]
Anything you change in `tpcccommon-1.0.tm` will affect the TPROC-C build and workload for all databases (not just Oracle you are working with). Likewise, anything in `tpchcommon-1.0.tm` will affect all TPROC-H.  Changing the local dbprefixoltp.tcl and dbprefixolap.tcl will only affect that particular workload for that particular database.

**Reply by vbienf** [reply]
Thanks! I think now I have all information that I need. I will update this topic after finish and testing with this files.

**Comment by vbienf** [comment]
Hi @sm-shaw !

I finished with first version. I changed scripts and now they generate compressible data like in the sysbench-tpcc ( Compression factor near 3.5 ) for MySQL. Next step it will be change tcl script for Oracle and Postgres.

Can you please check my git-repository ( https://github.com/vbienf/TPC-C-compressions-for-HammerDB ) and review this changes? I need your approve  that I changed only data generation and no more.

**Reply by sm-shaw** [reply]
To set expectations, I don't have the bandwidth to review right now. However, note the workload itself is in stored procedures, so if you changed the data only you should be good to go.

**Reply by vbienf** [reply]
Nope, I didn't touch stored procedures. 
I created new functions in `tpcccommon-1.0.tm` for generate strings with prefix and compressing, like it was at sysbench:
- MakePrefZip 
- MakePrefAddress 
- MakePrefName 
- MakeCompressString 
- MakeItemString

And in  dbprefixoltp.tcl-file (now only in mysql - `mysqloltp.tcl`) I just call these functions instead of "MakeAlphaString". F.e. for customer table:

`#set c_first [ MakeAlphaString 8 16 $globArray $chalen ] `
`set c_first [ MakePrefName "first-" 2 10 $globArray $chalen ]`

Data generation that was changed:
1. Table warehouse
- w_name = name-<random string(5)>
- w_street_1 = street1-<random string(10)>
- w_street_2 = street2-<random string(10)>
- w_city = city-<random string(10)>
- w_zip = zip-<random number(5)>
2. Table district
- d_name = name-<random string(5)>
- d_street_1 = street1-<random string(10)>
- d_street_2 = street2-<random string(10)>
- d_city = city-<random string(10)>
- d_zip = zip-<random number(5)>
3. Table customer
- c_first = first-<random string(2,10)>
- c_street_1 = street1-<random string(10)>
- c_street_2 = street2-<random string(10)>
- c_city = city-<random string(10)>
- c_zip = zip-<random number(5)>
- c_data = <1 random char string(300,500)>
4. Table history
- h_data = <1 random char string(12,24)>
5. Table item
- i_name = item-<i_im_id>-<i_price>-<random string(5)>
- i_data = data-<i_name>-<random string(8)>
6. Table stock
- s_dist_01 = <1 random char string(24)>
- s_dist_02 = <1 random char string(24)>
- s_dist_03 = <1 random char string(24)>
- s_dist_04 = <1 random char string(24)>
- s_dist_05 = <1 random char string(24)>
- s_dist_06 = <1 random char string(24)>
- s_dist_07 = <1 random char string(24)>
- s_dist_08 = <1 random char string(24)>
- s_dist_09 = <1 random char string(24)>
- s_dist_10 = <1 random char string(24)>
- s_data = <1 random char string(26,50)>
7. Table order_line
- ol_dist_info = <1 random char string(24)>

**Comment by vbienf** [comment]
Hi @sm-shaw !

I noticed that version 5.0 of HammerDB was released, and I downloaded it from the website. However, I couldn’t find the “modules” and “src” folders that I used previously to modify the code for generating compressible data for my databases. It seems that in this version, the files are already compiled, and I’m not sure how to modify the sources for my task.

I also tried running ./hammerdb without compilation, but my Ubuntu 24 system only includes Tcl 8.6, not version 9. I then tried compiling HammerDB from source following the instructions on " https://www.hammerdb.com/docs/ch01s13.html ", but I couldn’t figure out how to install the databases and libraries into the /opt/ directory.

Could you please advise me on the best way to continue my work with HammerDB?

**Reply by sm-shaw** [reply]
Hi, yes HammerDB v5.0 (and all future HammerDB releases) are built as executables. You cannot use HammerDB from a standard Tcl binary, either 8.6 or 9. The correct approach is to build from source, then you will get both the release file builds and the Dev version that is the same as previous versions with all of the source files. 
For the build, the databases don't need to be in the /opt directory, they can be anywhere, the key step is to export the environment variables to the location of the respective config commands than can be used to find the location of the libraries. 
`export MYSQL_CONFIG=/opt/mysql-8.4.4-linux-glibc2.28-x86_64/bin`
Also note that if there are specific databases you don't want to install and use then comment out the line in the Setup file 
https://github.com/TPC-Council/HammerDB/blob/master/Build/Bawt-2.1.0/Setup/HammerDB-Linux.bawt
This then won't build the interface for that database e.g. the following to exclude Db2 for your build.
`#Setup db2tcl             db2tcl-2.0.1.1.7z              db2tcl.bawt	`
Once the build is  complete using `$ ./Build-Linux.sh x64 Setup/HammerDB-Linux.bawt update` in the release directory you will have the developers version of HammerDB that you can modify.

**Reply by vbienf** [reply]
oh, great! Thank you for your answer! I will try to exclude databases (like maria and db2) from my build and compile without them

**Reply by vbienf** [reply]
@sm-shaw sorry for bothering you, but it looks that something is wrong on download's files. I got an next error in log when trying to compile:
Error list:
  Hash keys of file /home/pliops/workspace/awide/HammerDB/Build/Bawt-2.1.0/InputLibs/tclpy-0.4.1.7z differ.
It downloads some file with hash and 7z archive but it looks that hash keys not valid

**Reply by sm-shaw** [reply]
Hi, this was a known issue, that should have been fixed last week. Please delete the files (or at least tclpy-0.4.1.7z) from InputLibs - they get downloaded as part of the build process,  also delete the Bawt-Build directory under Bawt-2.1.0. Then if you re-run the files should match the hashkeys and it will work.

**Reply by sm-shaw** [reply]
So from InputLibs, remove the files from here your local build directory `/home/pliops/workspace/awide/HammerDB/Build/Bawt-2.1.0/InputLibs`

**Reply by vbienf** [reply]
@sm-shaw I deleted all files from InputLibs folder and BawtBuild dirrectory, Then executed again "Build-Linux.sh"  but I got the same error with "Hash keys differ".

**Reply by sm-shaw** [reply]
I have doubled checked and re-uploaded the file, please try the process again.

**Reply by vbienf** [reply]
@sm-shaw now it starting to compile, but it seems that still some problem with **tclpy**:


12:14:56 > Summary
           Setup file     : /home/pliops/workspace/awide/HammerDB-5.0-git/Build/Bawt-2.1.0/Setup/HammerDB-Linux.bawt
           Build directory: /home/pliops/workspace/awide/HammerDB-5.0-git/Build/BawtBuild/Linux/x64/Release/Build
           Architecture   : x64
           Compilers      : gcc
           Global stages  : Finalize
           #  : Library Name         Version    Build time      Stages
           ----------------------------------------------------------------------
             1: Tcl                  9.0.1      2.26 minutes    Clean Extract Configure Compile Distribute
             2: TclS                 9.0.1      2.26 minutes    Clean Extract Configure Compile Distribute
             3: Tk                   9.0.1      0.09 minutes    Clean Extract Configure Compile Distribute
             4: TkS                  9.0.1      0.09 minutes    Clean Extract Configure Compile Distribute
             5: awthemes             10.4.0     0.00 minutes    Clean Extract Configure Compile Distribute
             6: expect               5.45.4.1   0.07 minutes    Clean Extract Configure Compile Distribute
             7: mysqltcl             3.052.1    0.03 minutes    Clean Extract Configure Compile Distribute
             8: oratcl               4.6.1      0.03 minutes    Clean Extract Configure Compile Distribute
             9: pgtcl                2.1.1.1    0.04 minutes    Clean Extract Configure Compile Distribute
            **10: tclpy                0.4.1      Excluded**        
            11: tclpyS               0.4.1      Excluded        
            12: ticklecharts         3.1        Excluded        
            13: tkblt                3.2.24     Excluded        
            14: tkpath               0.3.3.1    Excluded        
           ----------------------------------------------------------------------
           Total: 4.86 minutes
           
Error list:
  MSysRun: Error running command:
cd /home/pliops/workspace/awide/HammerDB-5.0-git/Build/BawtBuild/Linux/x64/Release/Build/**tclpy** ; make -j 64 ; make default 
Error message:
/home/pliops/workspace/awide/HammerDB-5.0-git/Build/BawtBuild/Linux/x64/Release/Build/tclpy /home/pliops/workspace/awide/HammerDB-5.0-git/Build/BawtBuild/Linux/x64/Release/Build/tclpy
sed "s/PACKAGE_VERSION/0.4.1/g" pkgIndex.tcl.in > pkgIndex.tcl
gcc -fPIC -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -D_FORTIFY_SOURCE=2 -Wl,-z,relro,-z,now -fno-strict-overflow -Wall -Wextra -Wstrict-aliasing -Wstrict-overflow -Wstrict-prototypes -Wno-unused-parameter -O2 -DPY_LIBFILE='"libpython3.8.a"' -DPACKAGE_VERSION='"0.4.1"' -I/usr/local/include/python3.8 -I/usr/local/include/python3.8 -I/home/pliops/workspace/awide/HammerDB-5.0-git/Build/BawtBuild/Linux/x64/Release/Install/Tcl/include -c generic/tclpy.c -o tclpy.o
rm -f libtclpy.so tclpy.so
gcc -shared -fPIC -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -D_FORTIFY_SOURCE=2 -Wl,-z,relro,-z,now -fno-strict-overflow -Wall -Wextra -Wstrict-aliasing -Wstrict-overflow -Wstrict-prototypes -Wno-unused-parameter -O2 -DPY_LIBFILE='"libpython3.8.a"' tclpy.o -o libtclpy0.4.1.so -Wl,--export-dynamic -L/home/pliops/workspace/awide/HammerDB-5.0-git/Build/BawtBuild/Linux/x64/Release/Install/Tcl/lib -ltcl9.0 -lpython3.8 -lcrypt -ldl  -lm -lm 
rm -f libtclpy.so tclpy.so
gcc -shared -fPIC -fstack-protector --param=ssp-buffer-size=4 -Wformat -Werror=format-security -D_FORTIFY_SOURCE=2 -Wl,-z,relro,-z,now -fno-strict-overflow -Wall -Wextra -Wstrict-aliasing -Wstrict-overflow -Wstrict-prototypes -Wno-unused-parameter -O2 -DPY_LIBFILE='"libpython3.8.a"' tclpy.o -o libtclpy0.4.1.so -Wl,--export-dynamic -L/home/pliops/workspace/awide/HammerDB-5.0-git/Build/BawtBuild/Linux/x64/Release/Install/Tcl/lib -ltcl9.0 -lpython3.8 -lcrypt -ldl  -lm -lm 
/home/pliops/workspace/awide/HammerDB-5.0-git/Build/BawtBuild/Linux/x64/Release/Build/tclpy
generic/tclpy.c:480:19: warning: cast between incompatible function types from ג€˜PyObject * (*)(PyObject *, PyObject *, PyObject *)ג€™ {aka ג€˜struct _object * (*)(struct _object *, struct _object *, struct _object *)ג€™} to ג€˜PyObject * (*)(PyObject *, PyObject *)ג€™ {aka ג€˜struct _object * (*)(struct _object *, struct _object *)ג€™} [-Wcast-function-type]
  480 |         {"eval",  (PyCFunction)tclpy_eval,
      |                   ^
/usr/bin/ld: /usr/local/lib/libpython3.8.a(object.o): warning: relocation against `PyDictKeys_Type' in read-only section `.text'
/usr/bin/ld: /usr/local/lib/libpython3.8.a(floatobject.o): relocation R_X86_64_PC32 against symbol `PyFloat_Type' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
make: *** [Makefile:56: libtclpy0.4.1.so] Error 1
/usr/bin/ld: /usr/local/lib/libpython3.8.a(object.o): warning: relocation against `PyDictKeys_Type' in read-only section `.text'
/usr/bin/ld: /usr/local/lib/libpython3.8.a(floatobject.o): relocation R_X86_64_PC32 against symbol `PyFloat_Type' can not be used when making a shared object; recompile with -fPIC
/usr/bin/ld: final link failed: bad value
collect2: error: ld returned 1 exit status
make: *** [Makefile:56: libtclpy0.4.1.so] Error 1
    while executing
"exec /bin/bash /home/pliops/workspace/awide/HammerDB-5.0-git/Build/BawtBuild/Linux/x64/Release/Build/tclpy/_Bawt_MSysRun_MSysBuild.sh"
    ("eval" body line 1)
    invoked from within
"eval exec [GetMSysShell] [list $batchFile]"

**Reply by sm-shaw** [reply]
Hi, it looks like another project hit the identical issue here https://github.com/gnuradio/gnuradio/issues/4783 this suggests that the issue is with linking to a custom Python build, I use the system installed ones.
